src/clients/FtpClient.cc

```diff
@@ -69,6 +69,61 @@ escapeIAC(const char *buf)
     return ret;
 }
 
+bool
+parseEPSV(Parser::Tokenizer &tz, const SBuf &tuple, uint16_t &outPort)
+{
+    try {
+        // '(' or need more data
+        tz.skipRequired("opening '('", SBuf("("));
+
+        // Peek the delimiter from the underlying buffer (tuple[1])
+        if (tuple.length() < 2)
+            throw TextException("empty delimiter in EPSV tuple", Here());
+
+        const auto delim = static_cast<unsigned char>(tuple[1]);
+
+        // validate delimiter: ASCII 33..126 and not a digit
+        if (delim < 33 || delim > 126 || xisdigit(delim))
+            throw TextException("invalid EPSV delimiter (must be ASCII 33..126 and not a digit)", Here());
+
+        // same delimiter three times
+        if (!tz.skip(static_cast<char>(delim)) ||
+                !tz.skip(static_cast<char>(delim)) ||
+                !tz.skip(static_cast<char>(delim)))
+            throw TextException("EPSV tuple missing repeated delimiters", Here());
+
+        // 1..5 decimal digits (port)
+        const auto port64 = tz.udec64("EPSV port", 5);
+
+        // 4th delimiter and ')'
+        if (!tz.skip(static_cast<char>(delim)))
```

Comment:
When the `xisdigit` and related checks are replaced by a `CharacterSet` test all these `static_cast<char>(delim)` can polish out.
```suggestion
        if (!tz.skip(delim) || !tz.skip(delim) || !tz.skip(delim))
            throw TextException("EPSV tuple missing repeated delimiters", Here());

        // 1..5 decimal digits (port)
        const auto port64 = tz.udec64("EPSV port", 5);

        // 4th delimiter and ')'
        if (!tz.skip(delim))
```